lib/src/oauth2_flows/auth_code.dart

```diff
@@ -136,6 +136,8 @@ abstract class AuthorizationCodeGrantAbstractFlow {
       'client_id=${Uri.encodeQueryComponent(clientId.identifier)}',
       'redirect_uri=${Uri.encodeQueryComponent(redirectUri)}',
       'scope=${Uri.encodeQueryComponent(scopes.join(' '))}',
+      'prompt=consent',
+      'access_type=offline',
```

Comment:
Can you elaborate on why this is correct?

Please reference documentation for the oauth2 end-point..

Also is this always correct in all the ways in which this method can be called?

I'm not convinced that all existing applications using this want an offline access token. I'll admit I haven't fully thought this through.. My inclination is that there is many contexts in which it is hard to use the `googleapis_auth` package.

Reply:
First of all, the need to configure with those parameters in the authentication of the _authenticationUri method, is to be able to obtain the necessary refresh_token in the request obtainAccessCredentialsUsingCode. Otherwise assert will invalidate continue flow in AutoRefreshingClient of auth_http_utils.dart

```
AutoRefreshingClient(Client client, this.clientId, this.credentials,
      {bool closeUnderlyingClient: false})
      : super(client, closeUnderlyingClient: closeUnderlyingClient) {
    assert(credentials.refreshToken != null);
    authClient = authenticatedClient(baseClient, credentials);
  }
```


We can see how to get access with this guide
https://developers.google.com/android-publisher/authorization#generating_a_refresh_token
or https://developers.google.com/oauthplayground/

This is the documentation that explains the purpose of the parameters.
https://developers.google.com/identity/protocols/oauth2/web-server#obtainingaccesstokens
access_type: Recommended
Indicates whether your application can refresh access tokens when the user is not present at the browser. Valid parameter values ​​are online, which is the default value, and offline.

Set the value to offline if your application needs to refresh access tokens when the user is not present at the browser. This is the method of refreshing access tokens described later in this document. This value instructs the Google authorization server to return a refresh token and an access token the first time that your application exchanges an authorization code for tokens.

prompt: consent Prompt the user for consent.

Now for what purpose would refresh tokens be used?
The one I am currently implementing is to be able to generate a backup process at a time where I do not need user assistance again for this action.

Reply:
Note: Please use [markdown quoting](https://docs.github.com/en/github/writing-on-github/basic-writing-and-formatting-syntax#quoting-text) when quoting text, this makes reading your response a lot easier.

> prompt: consent Prompt the user for consent.

Why are we making this change? As I understand [the docs](https://developers.google.com/identity/protocols/oauth2/web-server#creatingclient) this forces a prompt.

Indeed the default may be attractive in some cases:
> A space-delimited, case-sensitive list of prompts to present the user. If you don't specify this parameter, the user will be prompted only the first time your project requests access.

-----------------------------------
> I'm not convinced that all existing applications using this want an offline access token. I'll admit I haven't fully thought this through..

Someone needs to investigate this.. If we don't know if an offline access token is desired in **all existing** usages of the API, then **this is a breaking change**.

> First of all, the need to configure with those parameters in the authentication of the _authenticationUri method, is to be able to obtain the necessary refresh_token in the request obtainAccessCredentialsUsingCode. Otherwise assert will invalidate continue flow in AutoRefreshingClient of auth_http_utils.dart

I okay, I agree, that if the end result of function is to create `AutoRefreshingClient` then the lack of a `refresh_token` is probably going to cause said function to never work. So any exposed function that calls `AuthorizationCodeGrantAbstractFlow._authenticationUri` and then creates a `AutoRefreshingClient` will never have worked. And hence, fixing such functions is not a breaking change :)

But `obtainAccessCredentialsViaUserConsent` which is exposed from `lib/auth_io.dart` is not broken in the current state. It does not request a `refresh_token` and we cannot assume that all callers of [`obtainAccessCredentialsViaUserConsent`](https://pub.dev/documentation/googleapis_auth/latest/googleapis_auth.auth_io/obtainAccessCredentialsViaUserConsent.html) desires a flow which requests `refresh_tokens`. Okay, maybe we can, as I said I haven't research this to end. Maybe it's possible that for some reason `obtainAccessCredentialsViaUserConsent` would always want a `refresh_token`, but someone needs to go through the entire code base and make these arguments for every public function affected by this change.

Or maybe we have to change the internal logic to only make `refresh_tokens` when invoked in certain ways.

It's also possible that this can't be easily fixed and the best solution is to fork.